streams/lms_metadata.py

```diff
@@ -0,0 +1,180 @@
+class MetadataHolder:
+  """Contains metadata"""
+  album: str
+  artist: str
+  track: str
+  image_url: str
+
+  def print_meta(self):
+    """Prints contents of metadata"""
+    # Encased in line breaks because when it is being constantly printed, it visually mucks up the console a lot
+    logging.info(f"\nAlbum: {self.album}\nArtist: {self.artist}\nTrack: {self.track}\nImage: {self.image_url}\n")
+
+  def save_file(self, player):
+    data = {
+      'album': self.album,
+      'artist': self.artist,
+      'track': self.track,
+      'image_url': self.image_url
+    }
+    player_name = str(player).replace(' ', '_')
+
+    with open(f"lms_{player_name}_metadata_temp.json", 'wt', encoding='utf-8') as f:
+      json.dump(data, f, indent = 2)
+    os.replace(f"lms_{player_name}_metadata_temp.json", f"lms_{player_name}_metadata.json")
+
+
+class LMSMetadataReader:
+  """A class for getting metadata from a Logitech Media Server."""
+
+  # meta_ref is probably an unneccessary variable to pass as an arg since it's obscured from the user, but we can eventually make it an optional setting for the user
+  def __init__(self, name: str, server: Optional[str] = None, port: Optional[int] = 9000, meta_ref: Optional[int] = 2, debug: Optional[bool] = False):
+    self.player_name = name
+    self.server = server
+    self.port = port
+    self.ip = None # populated later
+    self.meta_ref_rate = meta_ref
+    self.debug = debug
+    self.meta = MetadataHolder(artist = 'Loading...',
+                          album = 'Loading...',
+                          track = 'Loading...',
+                          image_url = 'static/imgs/lms.png')
+    if not server:
+      self.meta.album =  'If loading takes a long time,'
+      self.meta.track = 'consider adding hostname to stream config'
+
+
+  def flatten(self, lms_info: dict) -> dict:
+    """ LMS returns data in a very verbose piecewise format to avoid name collisions,
+          this makes it easier to use, by disregarding possible collisions"""
+    x = 0
+    flat_data = {}
+    for item in lms_info:
+      for info in item:
+        flat_data[info] = lms_info[x][info]
+      x += 1
+    return flat_data
+
+  def connect(self):
+    """Discovers LMS Player and then requests metadata repetitively"""
+    connected = False
+    self.meta.save_file(self.player_name)
+
+    # When not connected, search for player to connect to by the proper name
+    while not connected:
+      try:
+        if self.server is not None:
+          self.ip = socket.gethostbyname(self.server)
+        else:
+          find_lms_server = "bin/arm/find_lms_server"
+          if platform.machine() == "x86_64":
+            find_lms_server = "bin/x64/find_lms_server"
+          # Much faster method of connecting to the metadata server using code from: https://github.com/ralph-irving/squeezelite/blob/master/tools/find_server.c
+          ip_find = subprocess.run([find_lms_server], check=True, capture_output=True, text=True)
```

Comment:
i'd love to see this more effectively case-statemented out; I can see a world where we run both x86_64, arm, and arm64.

is there a fallback for when these executables cannot be run? it doesn't appear we actually handle the exception potentially generated from `check=true`; that feels important here.

Reply:
Now handled explicitly with `except subprocess.CalledProcessError` in recent commit